torch/csrc/jit/script/exit_transforms.cpp

```diff
@@ -0,0 +1,394 @@
+    while (b->outputs().size() > 0) {
+      b->eraseOutput(0);
+    }
+  }
+
+  static void addIfOutputs(
+      Node* n,
+      at::ArrayRef<Value*> true_outs,
+      at::ArrayRef<Value*> false_outs) {
+    IfView if_view(n);
+    registerBlockOutputs(if_view.thenBlock(), true_outs);
+    registerBlockOutputs(if_view.elseBlock(), false_outs);
+    for (size_t i = 0; i < true_outs.size(); ++i) {
+      auto out_type =
+          unifyTypes(true_outs.at(i)->type(), false_outs.at(i)->type());
+      n->addOutput()->setType(*out_type);
+    }
+  }
+
+  ExitPair handleLoop(Node* node) {
+    auto loop_block = node->blocks().at(0);
+    auto exit_pair = transformExits(loop_block);
+
+    auto status = getStatus(exit_pair);
+    // once we transform returns, this will no longer be true
+    TORCH_INTERNAL_ASSERT(status == WILL);
+    registerBlockOutputs(loop_block, exit_pair.exitValues());
+    return exit_pair;
+  }
+
+  // creates a vector of uninitialized values of the same type as the
+  // values_to_match
+  std::vector<Value*> matchValuesWithUnitialized(
+      at::ArrayRef<Value*> values_to_match) {
+    std::vector<Value*> match_values;
+    for (Value* val : values_to_match) {
+      match_values.push_back(getUnitValue(val->type()));
+    }
+    return match_values;
+  }
+
+  // Recurses on the if node and returns its return status
+  // If status != WONT_RETURN, sets the block_return_val and has returned val
+  // of its parent block before exit
+  ExitPair handleIf(Node* node) {
+    auto true_block = node->blocks().at(0);
+    auto false_block = node->blocks().at(1);
+
+    auto true_pair = transformExits(true_block);
+    auto false_pair = transformExits(false_block);
+    auto true_status = getStatus(true_pair);
+    auto false_status = getStatus(false_pair);
+
+    if (true_status == WONT && false_status == WONT) {
+      return ExitPair(false_val, std::vector<Value*>({}));
+    }
+
+    {
+      // for the block that is not exitting, its' exit values will not get
+      // used so we create uninitialized values of the same type as the other
+      // block
+      if (true_status == WONT) {
+        WithInsertPoint insert(true_block);
+        std::vector<Value*> exit_vals =
+            matchValuesWithUnitialized(false_pair.exitValues());
+        true_pair = ExitPair(false_val, exit_vals);
+      } else if (false_status == WONT) {
+        WithInsertPoint insert(false_block);
+        std::vector<Value*> exit_vals =
+            matchValuesWithUnitialized(true_pair.exitValues());
+        false_pair = ExitPair(false_val, exit_vals);
+      }
+    }
+    Value* has_exited;
+    if (true_status == WILL && false_status == WILL) {
+      // Need to maintain the invariant that
+      // true_val == WILL, false_val == WONT, else MIGHT
+      has_exited = true_val;
+    } else {
+      addIfOutputs(node, {true_pair.hasExited()}, {false_pair.hasExited()});
+      has_exited = node->outputs().at(node->outputs().size() - 1);
```

Comment:
Could we use the same code for the "true_status == WILL && false_status == WILL" code as well? I understand that there we can optimize out `has_exited` and not include it to the if outputs, but I think we are in a state when the implementation readability matter much more than this.

Reply:
It's commented above, we need to maintain the invariant that true_val == WILL, false_val == WONT, else might

Reply:
This is actually confusing: we have `true_val_` which is just `const=1` and we have a `true_status` which is an exit status of the true branch of the if. I think we need to rename it to avoid name conflict.

Also, maybe partially because of the what I said above, I don't understand what this comment means:
"Need to maintain the invariant that true_val_ == WILL, false_val_ == WONT, else MIGHT".

Reply:
renamed to then & else instead of true and false